### ISSUE TITLE
[FireMonkey] Tracking protection is not handled properly

### ISSUE THREAD

**Issue by qsniyg** [erosman/support#134]
For example, if you set Tracking Protection to strict, and use the following userscript:

```
// ==UserScript==
// @name test
// @include *
// @grant GM_xmlhttpRequest
// @grant GM.xmlHttpRequest
// @version        1.0
// ==/UserScript==



(function() {
  var data = {
    url: "https://pbs.twimg.com/media/EK3eZtaVAAAB256.jpg?name=orig",
    method: "GET",
    onload: function(resp) {
      console.log("onload", resp);
    },
    onerror: function(resp) {
      console.log("onerror", resp);
    }
  };

  try {
    GM_xmlhttpRequest(data);
  } catch (e) {
    GM.xmlHttpRequest(data);
  }

})();
```

It will return both `onload null` and `onerror null`, which both causes issues as `onload` and `onerror` are not supposed to be both called, and also because it should still return an object (with `status` set to 0).

Violentmonkey, Tampermonkey, and Greasemonkey instead return:

```
onerror {
context: undefined
finalUrl: "https://pbs.twimg.com/media/EK3eZtaVAAAB256.jpg?name=orig"
lengthComputable: false
loaded: 0
readyState: 4
response: ""
responseHeaders: ""
responseText: ""
status: 0 // Tampermonkey and Greasemonkey currently null for status, but this is a bug that will be fixed in the next Tampermonkey version, and hopefully in the next Greasemonkey version as well
statusText: ""
total: 0
}
```

**Comment by erosman** [comment]
I tested the following with BOTH **Strict** & **Custom** **Enhanced Tracking Protection** and the were no errors.


```
(function() {
  var data = {
    url: "https://github.com/erosman/support/issues/134",
    method: "GET",
    onload: function(resp) {
      console.log("onload", resp);
    },
    onerror: function(resp) {
      console.log("onerror", resp);
    }
  };

  try {
    GM_xmlhttpRequest(data);
  } catch (e) {
    GM.xmlHttpRequest(data);
  }

})();
```

Please note that in your code, you are getting a binary image `https://pbs.twimg.com/media/EK3eZtaVAAAB256.jpg?name=orig`

Anyway, the issue does not seem to be related to **Enhanced Tracking Protection**

> It will return both onload null and onerror null, which both causes issues as onload and onerror are not supposed to be both called

Let me investigate this further and find out the reason.

**Comment by qsniyg** [comment]
Is twitter or twimg set as an exception? With a new firefox profile, tracking protection blocks the link for me...

**Comment by erosman** [comment]
I also get the error logged to console as:

> The resource at “https://pbs.twimg.com/media/EK3eZtaVAAAB256.jpg?name=orig” was blocked because content blocking is enabled.    

I am working on the `null` issue at the moment.

**Comment by erosman** [comment]
the `null` issue is fixed. It now `onerror` returns (and no onload)
```
{
  "error": "error",
  "responseHeaders": "",
  "status": 0,
  "statusText": ""
}
```

AFA "`because content blocking is enabled`", I am not sure about this. It is a setting by the user and overriding or bypassing user-settings would be against user preference.

**Comment by qsniyg** [comment]
@erosman Thanks for fixing it! Is there a reason for the extra "error" field, and why `finalUrl` etc. is missing? It should only run the `onerror` callback...

**Comment by erosman** [comment]
> Is there a reason for the extra "error" field

I set it to process different errors ie: error/abort/timeout

It is not necessary as far as user-script development is concerned.

> why finalUrl etc. is missing?

Initially, I set the minimum necessary. I can add more property/values. Are they needed?

**Comment by qsniyg** [comment]
@erosman 

>Are they needed?

They could be, the reason why I asked is because that way, it matches the behavior of other userscript managers. Currently I have no use for them for this specific event (the URL is blocked), but others might.

**Comment by erosman** [comment]
OK.. I will make up a standard return for all of them

Do you have examples of return by each of GM/VM/TM?

**Comment by qsniyg** [comment]
VM:

`{"finalUrl":"https://pbs.twimg.com/media/EK3eZtaVAAAB256.jpg?name=orig","readyState":4,"responseHeaders":"","status":0,"statusText":"","responseText":"","response":""}`

GM:

`{"context":null,"finalUrl":"https://pbs.twimg.com/media/EK3eZtaVAAAB256.jpg?name=orig","lengthComputable":null,"loaded":null,"readyState":4,"response":"","responseHeaders":null,"responseText":"","responseXML":false,"status":null,"statusText":null,"total":null}`

TM:

`{"readyState":4,"responseHeaders":"","finalUrl":"https://pbs.twimg.com/media/EK3eZtaVAAAB256.jpg?name=orig","status":0,"statusText":"","responseXML":null,"responseText":"","response":null}`

Personally I'd say go for TM/VM's style (`0`/`""` instead of `null`), as they are the standard nowadays.

**Comment by erosman** [comment]
> Personally I'd say go for TM/VM's style (0/"" instead of null), as they are the standard nowadays.

I also set to `"status": 0`

**Comment by erosman** [comment]
Now FireMonkey returns the SAME full set on all onload/onerror/onabort/ontimeout
onerror example:

```
{
  "readyState": 4,
  "response": "",
  "responseHeaders": "",
  "responseText": "",
  "responseType": "",
  "responseURL": "https://pbs.twimg.com/media/EK3eZtaVAAAB256.jpg?name=orig",
  "responseXML": null,
  "status": 0,
  "statusText": "",
  "timeout": 0,
  "withCredentials": false,
  "finalUrl": "https://pbs.twimg.com/media/EK3eZtaVAAAB256.jpg?name=orig"
}
```

**Comment by erosman** [comment]
The bug regarding `onload null` and `onerror null` is fixed in v1.24

The issue of **Custom Enhanced Tracking Protection** relates to Firefox settings and shouldn't be handled by the extension.
